Added companion (pet) to player

joberthel opened the pull request:
So I had a lot of fun implementing that one. As a player you can now select a companion (cat or dog), which will follow you arround on the map.

https://user-images.githubusercontent.com/34104366/113457742-dd698600-9410-11eb-9045-85e02cad234b.mp4

The companion movement will not be send over the network. It is calulated locally based on the player movement. But the selected companion will be send to other players. So you can also see the companions of other players.

I added a companion selection scene, were you can choose your favourite pet or none if you really don't want one. This scene will be shown, if it wasn't shown before or when you navigate to it via the new menu entry.

![companion-scene](https://user-images.githubusercontent.com/34104366/113457090-2e787a80-940f-11eb-9b8f-072520ea9bff.png)

![companion-menu](https://user-images.githubusercontent.com/34104366/113457095-30423e00-940f-11eb-8d1d-0ff632860dc2.png)

pizkaz commented:
Really cool! As soon as #873 gets merged I will add touch capability to the "select companion" scene. :-)

Kharhamel commented:
This looks really good! I need to test it a bit but we should be able to merge this quickly.

pizkaz commented on front/src/Phaser/Login/CustomizeScene.ts:
```diff
@@ -115,7 +116,7 @@ export class CustomizeScene extends AbstractCharacterScene {
 
             this.scene.sleep(CustomizeSceneName);
-            gameManager.tryResumingGame(this, EnableCameraSceneName);
+            gameManager.tryResumingGame(this, localUserStore.wasCompanionSet() ? EnableCameraSceneName : SelectCompanionSceneName);
```
I suggest a compromise: Show it only if the user selected "design your own [character]".

joberthel replied:
I changed it, so that it is only accessible from the menu. Showing it only after the "design your own" scene also makes sense. But showing it to someone, who just want's to quickly jump into a meeting probably doesn't.

joberthel commented on front/src/Phaser/Companion/Companion.ts:
```diff
@@ -0,0 +1,223 @@
+
+            this.getBody().stop();
+        } else {
+            this.animationType = PlayerAnimationTypes.Walk;
```
I tried that function. The pets won't get slower, if they come nearer. It basically looks like the pets are orbiting the player.

https://user-images.githubusercontent.com/34104366/113753247-2a4da500-970e-11eb-93a2-c93a48f4095f.mp4

Kharhamel replied:
Yes they will keep accelerating toward the player until they get passed him they will need to accelerate in another direction. I checked the doc: https://photonstorm.github.io/phaser3-docs/Phaser.Physics.Arcade.ArcadePhysics.html#accelerateToObject
and it doesn't look like you can tell the object to stop accelerating passed a specific distance. Sorry for giving you a bad advice.

joberthel commented on front/src/Phaser/Login/SelectCharacterScene.ts:
```diff
@@ -145,7 +146,7 @@ export class SelectCharacterScene extends AbstractCharacterScene {
         if (this.selectedPlayer !== null) {
             gameManager.setCharacterLayers([this.selectedPlayer.texture.key]);
-            gameManager.tryResumingGame(this, EnableCameraSceneName);
+            gameManager.tryResumingGame(this, localUserStore.wasCompanionSet() ? EnableCameraSceneName : SelectCompanionSceneName);
```
I removed it.

joberthel commented on front/src/Phaser/Login/SelectCompanionScene.ts:
```diff
@@ -0,0 +1,250 @@
+                && ((this.selectedRectangleYPos * this.nbCharactersPerRow) + (this.selectedRectangleXPos + 1) + 1) <= this.companionModels.length
+            ) {
+                this.selectedRectangleXPos++;
+            }
```
I refactored the whole selection stuff. Should be a lot simpler now.

joberthel commented on front/src/Phaser/Login/SelectCompanionScene.ts:
```diff
@@ -0,0 +1,250 @@
+    private getCompanionIndex(): number {
+        const companion = localUserStore.getCompanion();
+
+        if (companion === null) {
```
I added a check in the getCompanio function.

joberthel commented on messages/protos/messages.proto:
```diff
@@ -36,6 +36,10 @@ message CharacterLayerMessage {
   string name = 2;
 }
 
+message Companion {
```
I renamed it.

joberthel commented on front/src/Phaser/Login/SelectCompanionScene.ts:
```diff
@@ -0,0 +1,250 @@
+    }
+
+    preload() {
+        lazyLoadAllResources(this.load).then(() => {
```
I changed it, so that it is basically like in the character selection.

joberthel commented on front/src/Phaser/Companion/Companion.ts:
```diff
@@ -0,0 +1,223 @@
+
+        this.companionName = name;
+
+        lazyLoadResource(this.scene.load, this.companionName)
```
I added a texturePromise like in the character class.

Jacques-Olivier-Farcy commented:
I love this feature !!

justafish commented:
Super cool! :cat: 

For some reason this is currently crashing when trying to exit from one map to the other, but only when the user has a companion selected. Everything is freezes and this is dumped into the browser console (this is on the workadventure-xce fork) -

```
No audio element loaded to unload AudioManager.ts:173:20
Uncaught TypeError: e[t] is undefined
    shutdown phaser.js:166333
    emit phaser.js:1928
    shutdown phaser.js:48939
    stop phaser.js:100378
    processQueue phaser.js:99453
    update phaser.js:99707
    step phaser.js:162214
    step phaser.js:88795
    t phaser.js:89065
    setTimeout handler*t phaser.js:89067
    setTimeout handler*t phaser.js:89067
    setTimeout handler*t phaser.js:89067
    setTimeout handler*t phaser.js:89067
    setTimeout handler*t phaser.js:89067
    setTimeout handler*t phaser.js:89067
    setTimeout handler*t phaser.js:89067
    setTimeout handler*t phaser.js:89067
    setTimeout handler*t phaser.js:89067
    setTimeout handler*t phaser.js:89067
    setTimeout handler*t phaser.js:89067
    setTimeout handler*t phaser.js:89067
    setTimeout handler*t phaser.js:89067
    setTimeout handler*t phaser.js:89067
    setTimeout handler*t phaser.js:89067
    setTimeout handler*t phaser.js:89067
    setTimeout handler*t phaser.js:89067
    setTimeout handler*t phaser.js:89067
    setTimeout handler*t phaser.js:89067
    setTimeout handler*t phaser.js:89067
    setTimeout handler*t phaser.js:89067
    setTimeout handler*t phaser.js:89067
    setTimeout handler*t phaser.js:89067
    setTimeout handler*t phaser.js:89067
    setTimeout handler*t phaser.js:89067
    setTimeout handler*t phaser.js:89067
    setTimeout handler*t phaser.js:89067
    setTimeout handler*t phaser.js:89067
    setTimeout handler*t phaser.js:89067
    setTimeout handler*t phaser.js:89067
    setTimeout handler*t phaser.js:89067
    setTimeout handler*t phaser.js:89067
    setTimeout handler*t phaser.js:89067
    setTimeout handler*t phaser.js:89067
    setTimeout handler*t phaser.js:89067
    setTimeout handler*t phaser.js:89067
    setTimeout handler*t phaser.js:89067
    setTimeout handler*t phaser.js:89067
    setTimeout handler*t phaser.js:89067
    setTimeout handler*t phaser.js:89067
    setTimeout handler*t phaser.js:89067
    setTimeout handler*t phaser.js:89067
    setTimeout handler*t phaser.js:89067
    setTimeout handler*t phaser.js:89067
    setTimeout handler*t phaser.js:89067
    setTimeout handler*t phaser.js:89067
    setTimeout handler*t phaser.js:89067
    setTimeout handler*t phaser.js:89067
    setTimeout handler*t phaser.js:89067
    setTimeout handler*t phaser.js:89067
    setTimeout handler*t phaser.js:89067
    setTimeout handler*t phaser.js:89067
    setTimeout handler*t phaser.js:89067
    setTimeout handler*t phaser.js:89067
    setTimeout handler*t phaser.js:89067
    setTimeout handler*t phaser.js:89067
    setTimeout handler*t phaser.js:89067
    setTimeout handler*t phaser.js:89067
    setTimeout handler*t phaser.js:89067
phaser.js:166333:12
```

pizkaz commented:
Thx @justafish, I was about to file the exact same issue. :-)

joberthel commented:
I tried reproducing that issue with the maps in the repository (Floor 0, Floor 1), but couldn't reproduce it. Do you have some more information, like which browser you are using or a link to a map?

justafish commented:
@joberthel try this map, the exit is at the bottom https://maps.drupalcon.lullabot.com/scavenger_hunt.json
![image](https://user-images.githubusercontent.com/345613/114073013-bc82b400-989a-11eb-910b-b595366a8fdf.png)

justafish commented:
What's strange is the exit into that map from a different one works fine :thinking:

joberthel commented:
I could narrow it down to the phaser update from 3.24.1 to 3.54.0. Probably something was changed in the destroy behaviour. Need to read the changelogs.

Kharhamel commented:
Yes, the update was pretty massive and had a few breaking changes, especialy when it comes to the pipelines API. I didn't remember changing anything in the audioManager when I tested the update though

Kharhamel reviewed (approved):
Good for me. I need to check with the other two before merging, but I have good hopes this shouldbe in the next release

pizkaz commented:
> Yes, the update was pretty massive and had a few breaking changes, especialy when it comes to the pipelines API. I didn't remember changing anything in the audioManager when I tested the update though

The message `No audio element loaded to unload AudioManager.ts:173:20` is unrelated. We get that any time we exit a map.

joberthel commented:
So I added a fix, which adds compatibility with the new phaser version. The main problem was, that I destroyed the companion inside of the character destroy function. And this sometimes leads to an issue if the whole scene gets destroyed. I don't really understand, why this is not a problem with the player name for example, which also gets detroyed inside of the character destroy function, but it should work now. I moved the companion destroy call into the GameScene.

justafish commented:
That fixed it, thank you!

pizkaz commented:
Confirmed. After a fresh rebase onto current upstream the issue is fixed in XCE as well. :+1:
